[expo-branch] First 3rd party library unimodule wrapper!

sjchmiela opened the pull request:
# Why

We would like to provide a way to users to optionally include `react-native-branch` in their standalone apps. Fixes https://github.com/expo/expo/issues/5132.

# How

I added support for React Native-specific unimodules to React Native adapter. Right now it works like this:

- unimodule `expo-branch` depends on `react-native`, so it can use its API. It's no longer a uni-platform module, more an optinmodule (I guess the name won't stick)
- we copy `react-native-branch` vendored module directly to the unimodule. It will compile, since we added necessary dependencies to the unimodule
- `BranchPackage < org.unimodules.Package` implements `ReactPackage` interface too, forwarding all the calls to the vendored `RNBranchPackage`.
- when the module registry provider creates a module registry from a list of `Packages`, it can filter out `ReactPackages` — and so it does, putting them into a local internal module (`ReactPackagesProvider`) which only purpose is to carry those to `ModuleRegistryAdapter` which is responsible for creating a list of React Native modules out of a module registry. The adapter grabs the `ReactPackagesProvider` and adds React Native modules to the list.

# Test Plan

I have confirmed that adding `BranchPackage` to a list of installed unimodules' packages creates `RNBranchModule` and exports to JS.

sjchmiela committed:
[tools] When updating Android Branch code, update expo-branch unimodule

sjchmiela committed:
[@unimodules/react-native-adapter] Add ReactPackages support to the adapter

sjchmiela committed:
[android] Do not unnecessarily override a method

sjchmiela committed:
[ios] Update react-native-branch to 3.1.1

sjchmiela committed:
[expo-branch] Add expo-branch

wkozyra95 commented on packages/expo-branch/unimodule.json:
```diff
@@ -0,0 +1,7 @@
+  "name": "expo-branch",
+  "platforms": ["android"],
+  "android": {
+    "packages": ["expo.modules.branch.BranchPackage"]
```
it's not required to specify package here
https://github.com/unimodules/react-native-unimodules/blob/master/gradle.groovy#L61

sjchmiela replied:
Apart from `BranchPackage < org.unimodules.Package` we also have `vendored.RNBranchPackage < com.facebook.react.Package` in the same module, which introduces a conflict. Both removing the first one or modifying the second one would require manual changes when updating the vendored code.

https://github.com/unimodules/react-native-unimodules/blob/09234cff13d179c74cf347136c04a477621c9fc7/gradle.groovy#L66-L68

esamelson reviewed (commented):
awesome, this is really exciting -- thanks for this! 😄 what's here looks good, just a couple of comments--
- would be nice to add some internal documentation around this, e.g. how to add a new module this way, how to keep the branch module up to date, anything extra that needs to be done in the release process, etc. i think our internal docs are on the sparser side and just want to make sure we're taking opportunities to improve them!
- could we use expo-branch in ios as well? i know we can't do optional modules yet on ios, but it would be nice not to have branch (and other third party modules) in two different places in the repo.

sjchmiela committed:
[expotools] Allow steps to opt out of updating pbxproj

sjchmiela committed:
[expotools] Update react-native-branch config

Updating `react-native-branch` via expotools will now update `expo-branch` unimodule

sjchmiela committed:
[expo-branch] Move RNBranch from Exponent to expo-branch

sjchmiela committed:
[ios] Move ExpoKit EXBranchManager to unimodule

Removed ExpoKit integration, EXBranchManager is setup like for a plain react-native-branch library

sjchmiela committed:
[ios] Add ExpoKit integration to Branch

sjchmiela committed:
[expo-branch] Add CocoaPods integration

Run `pod install`

sjchmiela committed:
[expotools] Make `expo-branch` unversionable

Since we only support it in standalone apps, which contain only unversioned code, we don't need to version it

sjchmiela committed:
[@unimodules/core] Favor subclasses of singletons

sjchmiela commented:
As @esamelson proposed, I moved Branch code from iOS `Exponent/Versioned` to `expo-branch`. This allowed me to create a full-fledged `expo-branch` unimodule, which cleanly integrates with properly configured `AppDelegate`s (using `UMAppDelegateWrapper`). It also allowed me to remove Expo-specific changes from `RNBranch` in favor of subclassing it in `EXScopedBranch` belonging to `ExpoKit` (bare apps will get unchanged `RNBranch` wired up with `expo-branch/EXBranchManager` which removes any need for extra setup steps).

`expo-branch` code is easily updated using `expotools`:
```
et update-module -m react-native-branch -p ${platform} -c ${commit}
```

sjchmiela committed:
[changelog] Update changelog entry
